ZShare/ViewModels/ExtensionViewModel.swift

```diff
@@ -737,42 +737,69 @@ final class ExtensionViewModel {
         DDLogInfo("ExtensionViewModel: downloaded unsupported attachment")
 
         // Remove downloaded file, it won't be used anymore
-        try? self.fileStorage.remove(file)
+        try? fileStorage.remove(file)
 
         guard (url.host ?? "").contains("sciencedirect") else {
-            self.state.attachmentState = .failed(.downloadedFileNotPdf)
+            state.attachmentState = .failed(.downloadedFileNotPdf)
             return
         }
 
         // Try loading the url in webview to bypass redirects
 
         DDLogInfo("ExtensionViewModel: detected sciencedirect, trying redirect")
 
-        self.state.attachmentState = .downloading(0)
-        self.state.retryCount += 1
-
-        self.getRedirectedPdfUrl(from: url) { [weak self] newUrl, newCookies, newUserAgent, newReferrer in
-            guard let self = self else { return }
+        state.attachmentState = .downloading(0)
+        state.retryCount += 1
 
-            if let newUrl = newUrl, newUrl != url && self.state.retryCount < 3 {
-                self.download(item: item, attachment: attachment, attachmentUrl: newUrl, to: file, cookies: newCookies, userAgent: newUserAgent, referrer: newReferrer)
+        getRedirectedPdfUrl(from: url, timeout: .seconds(2)) { [weak self] newUrl, newCookies, newUserAgent, newReferrer in
+            guard let self else { return }
+            if let newUrl, newUrl != url && state.retryCount < 3 {
+                download(item: item, attachment: attachment, attachmentUrl: newUrl, to: file, cookies: newCookies, userAgent: newUserAgent, referrer: newReferrer)
                 return
             }
 
-            // Didn't help, report failed PDF download
-            self.state.attachmentState = .failed(.downloadedFileNotPdf)
+            // Check if there is a captcha, to show the web view and load the URL once more.
+            let captchaLocator = ".challenge-form"
+            let javascript = "document.querySelector('\(captchaLocator)') !== null"
+            webView?.call(javascript: javascript)
+                .observe(on: MainScheduler.instance)
+                .subscribe(onSuccess: { [weak self] result in
+                    guard let self else { return }
```

Comment:
Looks good, but a little messy, would it be possible to use recursion here? Feel free to separate it into functions a bit more if possible. If you think it won't be much cleaner then we can leave this as is.

Reply:
Check https://github.com/zotero/zotero-ios/pull/868/commits/9b028ccefe430eaf56f65d50a09b5acd2edde338 for a more streamlined implementation.